File: api/src/main/java/com/avast/metrics/TimerPairImpl.java
package com.avast.metrics;

import com.avast.metrics.api.Timer;
import com.avast.metrics.api.TimerPair;

import java.time.Duration;
import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;

public class TimerPairImpl implements TimerPair {

    private final Timer successTimer;
    private final Timer failureTimer;

    public TimerPairImpl(Timer successTimer, Timer failureTimer) {
        this.successTimer = successTimer;
        this.failureTimer = failureTimer;
    }

    @Override
    public TimeContext start() {
        return new Context(successTimer.start(), failureTimer.start());
    }

    @Override
    public void update(Duration duration) {
        successTimer.update(duration);
    }

    @Override
    public void updateFailure(Duration duration) {
        failureTimer.update(duration);
    }

    public <T> T time(Callable<T> operation) throws Exception {
        Timer.TimeContext successContext = successTimer.start();
        Timer.TimeContext failureContext = failureTimer.start();
        try {
            T result = operation.call();
            successContext.stop();
            return result;
        } catch (Exception ex) {
            failureContext.stop();
            throw ex;
        }
    }


    @Override
    public <T> CompletableFuture<T> timeAsync(Callable<CompletableFuture<T>> operation, Executor executor) throws Exception {
        Timer.TimeContext successContext = successTimer.start();
        Timer.TimeContext failureContext = failureTimer.start();
        try {
            CompletableFuture<T> promise = new CompletableFuture<>();
            CompletableFuture<T> future = operation.call();
            future.handleAsync((success, failure) -> {
                if (failure == null) {
                    successContext.stop();
                    promise.complete(success);
                } else {
                    failureContext.stop();
                    promise.completeExceptionally(failure);
                }
                return null;
            }, executor);
            return promise;
        } catch (Exception ex) {
            failureContext.stop();
            throw ex;
        }
    }

    private static class Context implements TimeContext {

        private final Timer.TimeContext successContext;
        private final Timer.TimeContext failureContext;

        Context(Timer.TimeContext successContext, Timer.TimeContext failureContext) {
            this.successContext = successContext;
            this.failureContext = failureContext;
        }

        @Override
        public void stop() {
            successContext.stop();
        }

        @Override
        public void stopFailure() {
            failureContext.stop();
        }
    }
}

File: scala-api/src/main/scala/com/avast/metrics/scalaapi/impl/TimerPairImpl.scala
package com.avast.metrics.scalaapi.impl

import java.time.{Duration => JDuration}
import com.avast.metrics.scalaapi.{Timer, TimerPair}

import scala.concurrent.duration.Duration
import scala.concurrent.{ExecutionContext, Future}
import scala.util.control.NonFatal
import scala.util.{Failure, Success}

private class TimerPairImpl(success: Timer, failure: Timer) extends TimerPair {

  override def start(): TimeContext = {
    val succCtx = success.start()
    val failCtx = failure.start()

    new TimeContext {
      override def stop(): Unit = succCtx.stop()

      override def stopFailure(): Unit = failCtx.stop()
    }
  }

  override def update(duration: JDuration): Unit = success.update(duration)

  override def updateFailure(duration: JDuration): Unit = failure.update(duration)

  override def update(duration: Duration): Unit = success.update(duration)

  override def updateFailure(duration: Duration): Unit = failure.update(duration)

  override def time[A](block: => A): A = {
    val succCtx = success.start()
    val failCtx = failure.start()

    try {
      val a = block
      succCtx.stop()
      a
    } catch {
      case NonFatal(thr) =>
        failCtx.stop()
        throw thr
    }
  }

  override def time[A](future: => Future[A])(implicit ec: ExecutionContext): Future[A] = {
    val succCtx = success.start()
    val failCtx = failure.start()
    try {
      val a = future
      a andThen {
        case Success(_) =>
          succCtx.stop()
        case Failure(thr) =>
          failCtx.stop()
      }
    } catch {
      case NonFatal(thr) =>
        failCtx.stop()
        throw thr
    }
  }
}
